<a href="https://colab.research.google.com/github/Codeginner/machineLearning-task/blob/main/9th%20Week%20Tasks/NN%20on%20(2x%20%2B%202)%20function/Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Gilman Muslih Z

1103201075

NN on (2x + 2) function

In [12]:
# Import modul numpy untuk melakukan operasi aritmatika
import numpy as np

# Membuat array X sebagai input model
X = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], dtype=float)

# Membuat array y sebagai output yang diharapkan
y = np.array([4.0, 6.0, 8.0, 10.0, 12.0, 14.0], dtype=float)

In [13]:
import tensorflow as tf # Import library tensorflow ntuk pembuatan dan pelatihan model machine learning
from tensorflow import keras # Keras adalah antarmuka tingkat tinggi yang mempermudah pembuatan dan pelatihan model neural network

# Membuat model neural network dengan satu layer menggunakan TensorFlow dan Keras
model_1_layer = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

In [14]:
# Mengompilasi model dengan konfigurasi tertentu
model_1_layer.compile(optimizer='sgd', loss='mean_squared_error')

In [15]:
# Melakukan pelatihan model menggunakan metode fit dengan jumlah iterasi (epoch) selama pelatihan 50 epoch
model_1_layer.fit(X, y, epochs=50)

Epoch 1/50
1/1 [==============================] - 0s 393ms/step - loss: 113.8610
Epoch 2/50
1/1 [==============================] - 0s 16ms/step - loss: 52.8964
Epoch 3/50
1/1 [==============================] - 0s 13ms/step - loss: 24.6787
Epoch 4/50
1/1 [==============================] - 0s 13ms/step - loss: 11.6173
Epoch 5/50
1/1 [==============================] - 0s 11ms/step - loss: 5.5707
Epoch 6/50
1/1 [==============================] - 0s 11ms/step - loss: 2.7707
Epoch 7/50
1/1 [==============================] - 0s 11ms/step - loss: 1.4734
Epoch 8/50
1/1 [==============================] - 0s 8ms/step - loss: 0.8715
Epoch 9/50
1/1 [==============================] - 0s 6ms/step - loss: 0.5916
Epoch 10/50
1/1 [==============================] - 0s 12ms/step - loss: 0.4607
Epoch 11/50
1/1 [==============================] - 0s 8ms/step - loss: 0.3987
Epoch 12/50
1/1 [==============================] - 0s 10ms/step - loss: 0.3687
Epoch 13/50
1/1 [==============================] - 0s 10ms

Proses pelatihan ini akan mengoptimalkan parameter-model (koefisien dan bias) untuk mencocokkan pola dalam data, dengan tujuan mengurangi nilai fungsi loss yang telah ditentukan sebelumnya (dalam hal ini, menggunakan Mean Squared Error).

In [16]:
# Melakukan prediksi menggunakan model terlatih dengan input prediksi nilai 10.0
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 98ms/step
[[23.550913]]


In [17]:
model_1_layer.fit(X, y, epochs=100) # Melakukan pelatihan model menggunakan metode fit dengan jumlah iterasi (epoch) selama pelatihan 100 epoch

Epoch 1/100
1/1 [==============================] - 0s 9ms/step - loss: 0.2594
Epoch 2/100
1/1 [==============================] - 0s 8ms/step - loss: 0.2575
Epoch 3/100
1/1 [==============================] - 0s 9ms/step - loss: 0.2556
Epoch 4/100
1/1 [==============================] - 0s 8ms/step - loss: 0.2538
Epoch 5/100
1/1 [==============================] - 0s 7ms/step - loss: 0.2519
Epoch 6/100
1/1 [==============================] - 0s 9ms/step - loss: 0.2501
Epoch 7/100
1/1 [==============================] - 0s 8ms/step - loss: 0.2483
Epoch 8/100
1/1 [==============================] - 0s 8ms/step - loss: 0.2465
Epoch 9/100
1/1 [==============================] - 0s 9ms/step - loss: 0.2447
Epoch 10/100
1/1 [==============================] - 0s 9ms/step - loss: 0.2429
Epoch 11/100
1/1 [==============================] - 0s 8ms/step - loss: 0.2411
Epoch 12/100
1/1 [==============================] - 0s 10ms/step - loss: 0.2393
Epoch 13/100
1/1 [==============================] - 0s 9ms/s

Proses pelatihan ini akan melanjutkan peningkatan parameter-model untuk lebih meningkatkan kemampuan model dalam mencocokkan pola dalam data pelatihan. Jumlah epoch yang lebih besar bisa membantu model untuk terus mengoptimalkan parameter hingga batas yang diperlukan.

In [18]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 81ms/step
[[23.07598]]


Model yang telah dilatih memprediksi bahwa output yang diharapkan untuk input [10.0] adalah sekitar 23.18762. Dengan kata lain, model menghasilkan perkiraan output berdasarkan pola yang telah dikenal saat dilatih pada data sebelumnya.

In [19]:
# membuat model neural network dengan tiga layer
model_3_layer = tf.keras.Sequential([
    # Layer pertama (input layer) dengan 1 neuron, input_shape=[1] menunjukkan satu fitur input
    keras.layers.Dense(units=1, input_shape=[1]),

    # Layer kedua (hidden layer) dengan 8 neuron
    keras.layers.Dense(units=8),

    # Layer ketiga (output layer) dengan 1 neuron
    keras.layers.Dense(units=1)
])

In [20]:
# mengompilasi model neural network dengan konfigurasi tertentu
model_3_layer.compile(optimizer='sgd', loss='mean_squared_error')

In [21]:
# melatih model selama 100 epoch
model_3_layer.fit(X, y, epochs=100)

Epoch 1/100
1/1 [==============================] - 1s 701ms/step - loss: 96.8866
Epoch 2/100
1/1 [==============================] - 0s 15ms/step - loss: 4.2966
Epoch 3/100
1/1 [==============================] - 0s 12ms/step - loss: 18.9528
Epoch 4/100
1/1 [==============================] - 0s 18ms/step - loss: 24.1087
Epoch 5/100
1/1 [==============================] - 0s 16ms/step - loss: 75.2526
Epoch 6/100
1/1 [==============================] - 0s 14ms/step - loss: 43.5748
Epoch 7/100
1/1 [==============================] - 0s 10ms/step - loss: 0.6441
Epoch 8/100
1/1 [==============================] - 0s 14ms/step - loss: 1.6349
Epoch 9/100
1/1 [==============================] - 0s 15ms/step - loss: 5.7860
Epoch 10/100
1/1 [==============================] - 0s 13ms/step - loss: 11.1164
Epoch 11/100
1/1 [==============================] - 0s 11ms/step - loss: 38.8572
Epoch 12/100
1/1 [==============================] - 0s 10ms/step - loss: 0.1457
Epoch 13/100
1/1 [=======================

In [22]:
# Menampilkan hasil prediksi
print(model_3_layer.predict([10.0]))

1/1 [==============================] - 0s 76ms/step
[[25.321419]]


Output dari model model_3_layer setelah melalui proses pelatihan 100 epoch  untuk memprediksi output untuk input 10.0 adalah sekitar 25.321419.